bpo-45509: Check gzip headers for corrupted fields

rhpvorderman opened the pull request:
Check whether the COMMENT, NAME and HCRC fields are correct.

EDIT: Additionally, slightly increase performance for header checking for headers produced by gzip.compress or zlib.compress.
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-45509](https://bugs.python.org/issue45509) -->
https://bugs.python.org/issue45509
<!-- /issue-number -->

rhpvorderman committed:
Check gzip headers for corrupted fields

serhiy-storchaka commented on Lib/test/test_gzip.py:
```diff
@@ -562,6 +563,35 @@ def test_decompress(self):
+        ]
+        for header in truncated_headers:
+            with self.subTest(header=header):
+                self.assertRaises(EOFError, gzip._read_gzip_header,
```
It is better to only use public API in tests if possible. Does the test work with `gzip.decompress()`.

serhiy-storchaka commented on Lib/test/test_gzip.py:
```diff
@@ -562,6 +563,35 @@ def test_decompress(self):
+                  b"name\x00"
+                  b"comment\x00")
+        true_crc = zlib.crc32(header) & 0xFFFF
+        corrupted_crc = zlib.crc32(header, 50) & 0xFFFF
```
You can use something like `corrupted_crc = true_crc ^ 0xFFFF` (all bits inverted) to guarantee that `corrupted_crc != true_crc`.

rhpvorderman replied:
Great suggestion!

serhiy-storchaka commented on Lib/gzip.py:
```diff
@@ -430,29 +430,42 @@ def _read_gzip_header(fp):
+        true_crc = zlib.crc32(header_buffer.getvalue()) & 0xFFFF
+        if header_crc != true_crc:
+            raise BadGzipFile(f"Corrupted gzip header. Checksums do not "
+                              f"match: {true_crc} != {header_crc}")
```
Would not be more convenient to output the CRC values in hex? `:04x`

rhpvorderman replied:
Yes, this will make finding the actual header easier if this error is thrown. Done.

serhiy-storchaka commented on Lib/gzip.py:
```diff
@@ -430,29 +430,42 @@ def _read_gzip_header(fp):
         raise BadGzipFile('Not a gzipped file (%r)' % magic)
-
-    (method, flag, last_mtime) = struct.unpack("<BBIxx", _read_exact(fp, 8))
+    header_buffer = io.BytesIO()
```
Try to use a `bytearray` instead of `io.BytesIO`. Would it be faster? In what cases?

rhpvorderman replied:
Performance improvement was necessary (see bug report). So I made it better the following way:

- When the data is read from `fp` a bytes object is created. Instead of passing it anonymously to the functions that performs the check it is saved as a local variable.
- The FHCRC flag is checked and the result is saved as `fhcrc`, so the check runs only once.
- When `fhcrc` is truthy, a `header_parts` list is created. 
- Every time  `fp` is read and `fhcrc` is truthy the already created bytes objects are appended to the `header_parts` list. List appending has a very low overhead. It simply expands an array of PyObject pointers. Pointers that we already have as the named bytes objects.
- When the header should be checked, a `b"".join(header_parts)` is called to recreate the header.

This means for the most common use cases (no flags or only FNAME set) that only a few extra truthy checks are performed. This comes at a very low cost of 3% (see the bug report for results).

This is worth it since it enables correct truncation checking for FNAME and FCOMMENT as well as enabling checking of the header CRC when FHCRC is set.

rhpvorderman committed:
Minor performance tweaks to _read_gzip_header

Call the bool method and cache the result for faster truth checking.
Do not test for empty bytes but use "not magic" instead for faster truth checking.

rhpvorderman committed:
Optimize _read_gzip_header for the most common code paths

Those are:
+ Only FNAME set. (Created by gzip and python's GzipFile)
+ No flags set. (Created by gzip.compress and zlib.compress with wbits=31)

hauntsaninja committed:
Merge branch 'main' into bpo-45509

rhpvorderman committed:
Merge branch 'main' into bpo-45509

rhpvorderman committed:
Update test_gzip.py

Remove double import of zlib

serhiy-storchaka committed:
Merge branch 'main' into bpo-45509

serhiy-storchaka reviewed (commented):
Thank you for your patience @rhpvorderman. I am not very happy that the code becomes more complicated, but checking the CRC of the header if it is provided is the right thing. You have made considerable efforts to minimize the cost. I am inclined to accept this PR.

Yet one question. Did you consider to use a bytearray instead of a list to accumulate the header data?

rhpvorderman committed:
Factor out reading until null and use a bytearray

rhpvorderman committed:
Simplify comments. Better grouping of statements

rhpvorderman committed:
Fix error in method

serhiy-storchaka commented on Lib/gzip.py:
```diff
@@ -484,40 +484,63 @@ def _read_exact(fp, n):
 
+def _read_until_null(fp):
+    '''Read until the first encountered null byte in fp'''
+    result = io.BytesIO()
```
What if pass the bytearray buffer as argument to `_read_until_null()`? Would not it be faster?

rhpvorderman replied:
Yes. That makes more sense since it is a one-off function.

rhpvorderman committed:
Do not store header for common FNAME flag

rhpvorderman committed:
Reduce number of object creations
